### PR TITLE
Fixed #24720 -- Moved append slash checking in CommonMiddleware to process_response()

### PR DESCRIPTION
**Pull request by electricjay**
Addresses issue #24720 - using middleware.common to append slashes causes extra overhead to all requests that do not end in a slash

https://code.djangoproject.com/ticket/24720

To avoid resolving the url twice for all urls that do not end in a
slash - only check if we should add the slash if we receive a 404
response in *process_response* (if we are prepending www  it still does
the check in *process_request* to avoid multiple redirects)

Avoiding this double url resolution speeds up the affected requests by about 5%

This pull request exposes a bug(?!?) in the TestClient (see
*test_client_regress.tests.ContextTests.test_nested_requests*).  It
appears that returning a redirect from process_response behaves
differently than returning a redirect from process_request that leads
to extra context being added to the response.context when making 
nested calls.  Possibly related to the *template_rendered* signal (or not).

### CODE REVIEW

File: tests/middleware/tests.py [hunk 1]
```diff
@@ -39,6 +39,8 @@ def test_append_slash_have_slash(self):
         """
         request = self.rf.get('/slash/')
         self.assertEqual(CommonMiddleware().process_request(request), None)
+        response = HttpResponseNotFound
```
Comment (kezabelle) [inline tests/middleware/tests.py]: Missing parens `()`

Comment (electricjay) [inline tests/middleware/tests.py]: Good eye

**Comment by electricjay** [comment]
I have opened a ticket for the test client bug: https://code.djangoproject.com/ticket/24722

I have written a simple test case to show the test client bug: 
https://github.com/electricjay/django/commit/91146c4d58f0dd39b789e63d0854cca7f0adcaf8

github branch: https://github.com/electricjay/django/tree/test_client_nested_testcase

**Comment by electricjay** [comment]
After some research I have determined that what I thought was a bug in the test client is just a minor issue with the test **test_client_regress.tests.ContextTests.test_nested_requests**.  The triggering of the rewrite middleware appears to be accidental and not an intentional part of the test.

The extra context is added to the response because a 404 response is generated using a template when the original url lookup fails.  We may want to note this in the documentation, but I am not sure of the best place for it.

**Comment by timgraham** [comment]
This looks good. Here are some cosmetic edits: http://dpaste.com/329WCVC
A good place to note the backwards incompatible change you mentioned would be the 1.9 release notes.

**Comment by timgraham** [comment]
@electricjay, are you able to apply my patch, add release notes, and squash commits?

**Comment by timgraham** [comment]
merged in 434d309ef6dbecbfd2b322d3a1da78aa5cb05fa8, thanks!

**Pull request closed without merging**